Cpx SV PR series, part-8: detecting and interpreting complex structural variants with multiple chimeric alignment

SHuang-Broad opened the pull request:
finally here.

closing the main feature requests made in #2703 .

Need to wait for #3752 and #3770 .

Will generate a separate dummy PR about how to run the whole prototyping holistic SV interpretation tool.

Some small patches need to be applied after this PR, basically for dumpster diving into 
* ambiguous events
* events where assembly contigs clearly couldn't tell a complete picture

but all parts of the diving suit are all available after this PR.

SHuang-Broad commented:
known problem: doesn't generate a VCF yet, but a custom file format for demonstration purpose. 
Translating from the custom file format to annotated VCF records is no difficult task, if the reported format is agreed to.

SHuang-Broad commented:
Ready for review once bug-fixing PR #3871 is reviewed.

SHuang-Broad commented:
Ready for review once bug-fixing PR #3871 is reviewed.

cwhelan commented on src/main/java/org/broadinstitute/hellbender/tools/spark/sv/discovery/prototype/CpxVariantDetector.java:
What's the case where `SimpleInterval.contains()` doesn't work for this? Should we fix that method?

SHuang-Broad replied:
that's right, and now replaced with `SimpleInterval.contains()`

cwhelan commented on src/main/java/org/broadinstitute/hellbender/tools/spark/sv/discovery/prototype/CpxVariantDetector.java:
Again please provide a message explaining what's wrong with this situation.

SHuang-Broad replied:
expanded and documented

SHuang-Broad commented on src/main/java/org/broadinstitute/hellbender/tools/spark/sv/discovery/AlignmentInterval.java:
```diff
@@ -558,4 +563,45 @@ static Cigar softOrHardReclip(final Cigar cigar, final CigarOperator clipOperato
     }
+
+    /**
+     * @return  inclusive, 1-based boundaries on read to which this AlignmentInterval belongs,
```
done

SHuang-Broad commented on src/main/java/org/broadinstitute/hellbender/tools/spark/sv/discovery/AlignmentInterval.java:
```diff
@@ -558,4 +563,45 @@ static Cigar softOrHardReclip(final Cigar cigar, final CigarOperator clipOperato
+     *          if the given {@code otherRefSpan} overlaps the ref span of this alignment;
+     *          otherwise return a tuple 2 of (-1, -1)
+     */
+    public Tuple2<Integer, Integer> overlapBoundariesOnRead(final SimpleInterval otherRefSpan) {
```
done

SHuang-Broad commented on src/main/java/org/broadinstitute/hellbender/tools/spark/sv/discovery/prototype/ContigAlignmentsModifier.java:
```diff
@@ -65,6 +55,30 @@
+     *
+     * @return true if {@code one} should yield the homologous sequence to {@code two}.
+     */
+    static boolean homologyYieldingStrategy(final AlignmentInterval one, final AlignmentInterval two,
```
renamed to `yieldHomologousSequenceToAlignmentTwo()`

SHuang-Broad commented on src/main/java/org/broadinstitute/hellbender/tools/spark/sv/discovery/prototype/ContigAlignmentsModifier.java:
```diff
@@ -65,6 +55,30 @@
+                // jumpLandingRefLoc is for "that jump's landing reference location"
+                final int jumpStartRefLoc = one.referenceSpan.getEnd(),
+                          jumpLandingRefLoc = two.referenceSpan.getStart();
+                oneYieldToTwo = jumpStartRefLoc <= jumpLandingRefLoc == one.forwardStrand;
```
done

SHuang-Broad commented on src/main/java/org/broadinstitute/hellbender/tools/spark/sv/discovery/prototype/ContigAlignmentsModifier.java:
```diff
@@ -65,6 +55,30 @@
     }
 
+    /**
+     * Implementing homology-yielding strategy between two alignments {@code one} and {@code two}.
```
done

SHuang-Broad commented on src/main/java/org/broadinstitute/hellbender/tools/spark/sv/discovery/prototype/CpxVariantDetector.java:
not necessary when this is made to emit VCF; right now if I remove it, IDE says type is wrong.

SHuang-Broad replied:
the whole write to file will be removed in the final commit, now that we do emit VCF.

SHuang-Broad commented on src/main/java/org/broadinstitute/hellbender/tools/spark/sv/discovery/prototype/CpxVariantDetector.java:
temporary solution due to outputting custom file format. not necessary when outputting VCF

SHuang-Broad replied:
now removed by commit `9110a5aef8ad60ba30a032567c64b0e7e406e59f`

SHuang-Broad commented on src/main/java/org/broadinstitute/hellbender/tools/spark/sv/discovery/prototype/CpxVariantDetector.java:
will do

SHuang-Broad replied:
done in the latest commit `9110a5aef8ad60ba30a032567c64b0e7e406e59f`

SHuang-Broad commented on src/main/java/org/broadinstitute/hellbender/tools/spark/sv/discovery/prototype/CpxVariantDetector.java:
done

SHuang-Broad commented on src/main/java/org/broadinstitute/hellbender/tools/spark/sv/discovery/prototype/CpxVariantDetector.java:
changed as suggested, thanks for catching this!

SHuang-Broad commented on src/main/java/org/broadinstitute/hellbender/tools/spark/sv/discovery/prototype/CpxVariantDetector.java:
yep, and reworded

SHuang-Broad commented on src/main/java/org/broadinstitute/hellbender/tools/spark/sv/discovery/prototype/CpxVariantDetector.java:
"UINS" is definitely a better name.

SHuang-Broad commented on src/main/java/org/broadinstitute/hellbender/tools/spark/sv/discovery/prototype/CpxVariantDetector.java:
added link to the function that does the check

SHuang-Broad commented on src/main/java/org/broadinstitute/hellbender/tools/spark/sv/discovery/prototype/CpxVariantDetector.java:
found a bug here, and noted in a todo that will be done in the next logic change commit.

SHuang-Broad replied:
also moved the argument check into method `liveAbroad(...)`

SHuang-Broad commented on src/main/java/org/broadinstitute/hellbender/tools/spark/sv/discovery/prototype/CpxVariantDetector.java:
done, not sure if clear enough, please let me know if it is still unclear.

SHuang-Broad commented on src/main/java/org/broadinstitute/hellbender/tools/spark/sv/discovery/prototype/CpxVariantDetector.java:
done, not sure if clear enough, please let me know if it is still unclear.
Basically, there might be new edge cases that I haven't seen and 
my poor imagination is not catching them up, 
so this was for my own consumption when coded.

SHuang-Broad commented on src/main/java/org/broadinstitute/hellbender/tools/spark/sv/discovery/prototype/CpxVariantDetector.java:
renamed but I'm afraid I bad at naming stuff... :sweat:

SHuang-Broad commented on src/main/java/org/broadinstitute/hellbender/tools/spark/sv/discovery/prototype/CpxVariantDetector.java:
renamed as suggested.

SHuang-Broad commented on src/main/java/org/broadinstitute/hellbender/tools/spark/sv/discovery/prototype/CpxVariantDetector.java:
oversight, corrected

SHuang-Broad commented on src/main/java/org/broadinstitute/hellbender/tools/spark/sv/discovery/prototype/CpxVariantDetector.java:
I agree it is backwards. But...

The reason was that the (naive) alignment configuration scoring module rightnow uses MQ and AS (aligner score) for picking the "best" configuration (i.e. sub-list of the alignments given by aligner), which would be technically wrong if we were to split the gap and to simply grab the originating alignment's values.

This is especially true for AS, whose recomputing takes more time, and code, and forces us to know how AS are computed in the aligner so that there's no bias in computing the scores of naive alignments vs gap-split alignments (may not matter in practice, but still takes more code to compute).

Lots of the code in the discovery stage was devoted actually to alignment related acrobatics and edge cases so that the breakpoints we could resolve are as accurate as possible. 
I've kept in mind your wisdom that different aligners may be experimented with, but it seems unlikely in the near future (their own quirkiness, lack of API for JNI, etc); it seems more and more likely to me that eventually it's inevitable to have a custom alignment module in a high-quality SV pipeline, but again, the near future has other top priorities.

What do you think?

SHuang-Broad replied:
added a line of doc briefly explaining the reason

cwhelan commented:
If all we’re doing is splitting the alignment can’t we keep the same mapq and alignment score? We already know that the fragment that we split aligns to this location with those scores. The scores shouldn’t get worse just because we split the alignment (which is really just a formatting change). If the issue is that we need to sum over alignment scores, why not just split the score in half and give half to each side? We’re never going to select only one of the two halves of the split, right?

> On Dec 10, 2017, at 10:02 PM, Steve Huang <notifications@github.com> wrote:
> 
> @SHuang-Broad commented on this pull request.
> 
> In src/main/java/org/broadinstitute/hellbender/tools/spark/sv/discovery/prototype/CpxVariantDetector.java:
> 
> > +            this.tigWithInsMappings = new AssemblyContigWithFineTunedAlignments(contig, tigWithInsMappings.insertionMappings);
> +
> +            this.basicInfo = new BasicInfo(contig);
> +
> +            annotate(refSequenceDictionary);
> +        }
> +
> +        private static List<AlignmentInterval> deOverlapAlignments(final List<AlignmentInterval> originalAlignments,
> +                                                                   final SAMSequenceDictionary refSequenceDictionary) {
> +            final List<AlignmentInterval> result = new ArrayList<>(originalAlignments.size());
> +            final Iterator<AlignmentInterval> iterator = originalAlignments.iterator();
> +            AlignmentInterval one = iterator.next();
> +            while (iterator.hasNext()) {
> +                final AlignmentInterval two = iterator.next();
> +                // TODO: 11/5/17 an edge case is possible where the best configuration contains two alignments,
> +                //       one of which contains a large gap, and since the gap split happens after the configuration scoring,
> I agree it is backwards. But...
> 
> The reason was that the (naive) alignment configuration scoring module rightnow uses MQ and AS (aligner score) for picking the "best" configuration (i.e. sub-list of the alignments given by aligner), which would be technically wrong if we were to split the gap and to simply grab the originating alignment's values.
> 
> This is especially true for AS, whose recomputing takes more time, and code, and forces us to know how AS are computed in the aligner so that there's no bias in computing the scores of naive alignments vs gap-split alignments (may not matter in practice, but still takes more code to compute).
> 
> Lots of the code in the discovery stage was devoted actually to alignment related acrobatics and edge cases so that the breakpoints we could resolve are as accurate as possible.
> I've kept in mind your wisdom that different aligners may be experimented with, but it seems unlikely in the near future (their own quirkiness, lack of API for JNI, etc); it seems more and more likely to me that eventually it's inevitable to have a custom alignment module in a high-quality SV pipeline, but again, the near future has other top priorities.
> 
> What do you think?
> 
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub, or mute the thread.
>

SHuang-Broad commented on src/main/java/org/broadinstitute/hellbender/tools/spark/sv/discovery/prototype/CpxVariantDetector.java:
you are right, thanks! I became aware of this in the cosmetic-change commit, and should now be fixed.

SHuang-Broad commented:
>We already know that the fragment that we split aligns to this location with those scores. The scores shouldn’t get worse just because we split the alignment (which is really just a formatting change).

I am not sure I quite understand this comment. The (large) gapped alignment (aka the originating alignment), when split into child alignments, shouldn't have the same aligner score as each of the child, right? To me the AS is a score computed from similarities between the query and the target based on given penalties (at least most aligners does it this way), so if a large gap exists, the score actually takes that gap into account (opening and continuation). When we split the originating alignment into its children, each child alignment should have its own score, with their sum not the same as the originating alignment's because the gap is now not part of any child alignment.

>We’re never going to select only one of the two halves of the split, right?

Correct, but I am having thoughts about this now (not to pick only one&mdash;that would be wrong&mdash;but to ditch them altogether probably under some condition and redo the alignment step), exactly because of this behavior I observe. Think about the case where one originating gapped (say insertion) alignment, after splitting, has one of the two children contained in another alignment (not its sibling, that's impossible) in terms of their read span. Now the originating gapped alignment probably should be filtered out, or not, because if we keep it, an insertion would be called but apparently there are alternative explanations due to the other alignment. 
I'm not sure how to deal with this case, and if this scenario is common enough. It probably is the case that such alignments happen mostly in STR regions, so getting the exact alignments correct there is no easy task.

cwhelan commented:
>
> I am not sure I quite understand this comment. The (large) gapped
> alignment (aka the originating alignment), when split into child
> alignments, shouldn't have the same aligner score as each of the child,
> right? To me the AS is a score computed from similarities between the query
> and the target based on given penalties (at least most aligners does it
> this way), so if a large gap exists, the score actually takes that gap into
> account (opening and continuation). When we split the originating alignment
> into its children, each child alignment should have its own score, with
> their sum not the same as the originating alignment's because the gap is
> now not part of any child alignment.
>

That's true from the definition of aligner score, but for our
breakpoint-calling purposes, we're just reformatting the alignments so
they're in an easier format to work with. I.e. we could have kept the large
gap in the alignment, and then added some extra logic to call events based
on the cigar of the alignment, with no difference in the outcome. Just
because we split it to make things easier for us code-wise doesn't mean we
should have to recompute aligner scores -- we should just use the score of
the un-split alignment (since presumably that's a longer alignment with a
higher mapping quality).

But, maybe in the end I'm arguing for keeping the splitting after the
scoring, just so that we can preserve that original score and mapping
quality of the gapped alignment.

Correct, but I am having thoughts about this now (not to pick only one—that
> would be wrong—but to ditch them altogether probably under some condition
> and redo the alignment step), exactly because of this behavior I observe.
> Think about the case where one originating gapped (say insertion)
> alignment, after splitting, has one of the two children contained in
> another alignment (not its sibling, that's impossible) in terms of their
> read span. Now the originating gapped alignment probably should be filtered
> out, or not, because if we keep it, an insertion would be called but
> apparently there are alternative explanations due to the other alignment.
> I'm not sure how to deal with this case, and if this scenario is common
> enough. It probably is the case that such alignments happen mostly in STR
> regions, so getting the exact alignments correct there is no easy task.
>
Is that enough of a concern to worry about. In such a case I feel like we
should probably just pick the longer, gapped alignment, since it explains
more of the contig. But you have a better sense of how that fits in with
the rest of your scheme.

SHuang-Broad commented on src/main/java/org/broadinstitute/hellbender/tools/spark/sv/discovery/prototype/CpxVariantDetector.java:
the method is used in 3 different places, so I factored out a function with a clear name for it.
Do you think these short functions, though used several times, should be inlined?
If so, I'd inline them as I myself am struggling to make a decision.

SHuang-Broad commented:
Considering that this PR has lasted through my absence and the holiday season, I want to take the chance of summarizing the concerns you have issued here:

--------

Resolved as requested (or at least made efforts to):

 * documenting the logic and methods
   documented

 * emit VCF instead of custom file format
   emit both VCF custom file format now

 * bug in determining if alignment signature satisfies `allMiddleAlignmentsDisjointFromAlphaOmega`
   bug fixed in commit b4f7568b03b91eb77d256bcfe8117001bce040ec

--------

Unresolved yet:
the fact that gap split happens after the alignment configuration scoring step is considered backwards.

I agree in principle but due to AS and MQ were used in the scoring step, and split-copy leads to technically wrong AS & MQ, I originally decided to score first, then split.

Splitting the gapped alignments was introduced originally to have a centralized logic in inferring type and location of the events. 

The tension is that AS is used in the scoring but becomes practically useless after that.

>> Correct, but I am having thoughts about this now (not to pick only one—that
 would be wrong—but to ditch them altogether probably under some condition
 and redo the alignment step), exactly because of this behavior I observe.
 Think about the case where one originating gapped (say insertion)
 alignment, after splitting, has one of the two children contained in
 another alignment (not its sibling, that's impossible) in terms of their
 read span. Now the originating gapped alignment probably should be filtered
 out, or not, because if we keep it, an insertion would be called but
 apparently there are alternative explanations due to the other alignment.
 I'm not sure how to deal with this case, and if this scenario is common
 enough. It probably is the case that such alignments happen mostly in STR
 regions, so getting the exact alignments correct there is no easy task.
 
> Is that enough of a concern to worry about. In such a case I feel like we
should probably just pick the longer, gapped alignment, since it explains
more of the contig. But you have a better sense of how that fits in with
the rest of your scheme.

The comments I put in the code/todo was not clear (my bad). 
What the code is currently doing is what's suggested, that is: 
skipping the alignment that is BEFORE the child alignment from the gap-split, IFF that alignment contains the child alignment in terms of their spans on the read/contig; (I've updated the doc in the code as well). If you are concerned about the first child alignment from the same gapped alignment being skipped, don't worry, that is impossible because child alignments of the same gapped alignment cannot overlap on the read.

--------

Do these cover your major concerns @cwhelan?
Thanks!

cwhelan commented:
@SHuang-Broad thanks for the summary. I think this is OK to go in (once you resolve conflicts and make sure tests pass) -- we can go over and resolve some of the details in future changes.
